src/main/java/org/minions/devfund/noemiguzman/Squarelotron.java

```diff
@@ -0,0 +1,297 @@
+     * flip direction - upsideDown
+     *
+     * @param ring - ring number
+     * @return NewSquarelotron
+     */
+
+    public Squarelotron upsideDownFlip(int ring) {
+        // deep copy the board out
+        int[][] deepCopy1 = new int[this.size][this.size];
+        int[][] deepCopy2 = new int[this.size][this.size];
+        deepCopy1 = copySquarelotron(deepCopy1);
+        deepCopy2 = copySquarelotron(deepCopy2);
+
+        // manipulating on deepCopy2, upsideDownFlip on deepCopy2
+        for (int i = 0; i < this.size / DIMENSION; i++) {
+            int[] temp = deepCopy2[i];
+            deepCopy2[i] = deepCopy2[deepCopy2.length - i - 1];
+            deepCopy2[deepCopy2.length - i - 1] = temp;
+        }
+        // get the wanted ring on deepCopy2, and copy those numbers onto deepCopy1
+        // upper bound
+        deepCopy1 = rotateSquarelotron(ring, deepCopy1, deepCopy2);
+        return copyToNewSquarelotron(deepCopy1);
+    }
+
+    /**
+     * Copy from int[][] to new squareloton.
+     * @param deepCopy1  data to copy on new squareletorn
+     * @return Squarelotron
+     */
+    private Squarelotron copyToNewSquarelotron(final int[][] deepCopy1) {
+        Squarelotron s = new Squarelotron(this.size);
+        for (int i = 0; i < s.size; i++) {
+            for (int j = 0; j < s.size; j++) {
+                s.squarelotron[i][j] = deepCopy1[i][j];
+            }
+        }
+        return s;
+    }
+
+    /**
+     * inverse the squarelotron with the indicated ring number.
+     * inverse direction - transpose
+     * @param ring - ring number
+     * @return New Squarelotron
+     */
+    public Squarelotron mainDiagonalFlip(int ring) {
+        // deep copy the board out
+        int[][] deepCopy1 = new int[this.size][this.size];
+        int[][] deepCopy2 = new int[this.size][this.size];
+        deepCopy1 = copySquarelotron(deepCopy1);
+        deepCopy2 = copyReverseSquarelotron(deepCopy2);
+        deepCopy1 = rotateSquarelotron(ring, deepCopy1, deepCopy2);
+        // get the wanted ring on deepCopy2, and copy those numbers onto deepCopy1
+        // upper bound
+        // use deepCopy1 to get a new squarelotron
+
+        return copyToNewSquarelotron(deepCopy1);
+    }
+
+
+    /**
+     * rotate the squarelotron.
+     *
+     * @param numberOfTurns - indicating how it will turn
+     */
+
+    public void rotateRight(int numberOfTurns) {
+        // three different situation
+        // -3 or 1
+        // 1. diagonal flip (transpose) 2.upside down flip
+        // -2 or 2
+        // 1. left to right flip 2.upside down flip
+        // -1 or 3
+        // 1. main inverse diagonal flip 2.left to right flip
+        // 0 is no change at all
+        int ind = numberOfTurns % NUM_SIDES;
+        if (ind == -1 || ind == NUM_SIDES - 1) {
+            // transpose
+            int[][] temp = new int[this.size][this.size];
+            temp = copyReverseSquarelotron(temp);
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Remove this useless assignment; "temp" already holds the assigned value along all execution paths. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS4165)